crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1028,9 +1034,9 @@ impl Type {
             Type::Forall(typevars, typ) => {
                 let replacements = typevars
                     .iter()
-                    .map(|(id, var)| {
-                        let new = interner.next_type_variable();
-                        (*id, (var.clone(), new))
+                    .map(|(id, _)| {
+                        let new = interner.next_shared_type_variable();
+                        (*id, (new.clone(), Type::TypeVariable(new)))
```

Comment:
This is the other relevant portion and the main cause of the error. We should use `new` for both elements of the pair here rather than forwarding the old variable. Since the old variable is useles once we call `substitute` old for new replacements on the line after this.